### ISSUE TITLE
erased best_model problem

### ISSUE THREAD

**Issue by jerabaul29** [tensorforce/tensorforce#629]
@AlexKuhnle regarding the erasing of the best_model dump by the saver.

- we can avoid it by setting a high value of the saver frequency, as we have already discussed
- however, this has inconvenient of its own.

Would it be possible to mitigate this by copying in python (shutil.copy or something like that) the best_model to a separate folder one level lower in the saver tree, each time it is updated? Would it be a problem if this is defined as a pure python operation, rather than a tensorflow node?

**Comment by jerabaul29** [comment]
Ok, good to know.

I do it this way (may be old / not how to do it longer though):

```
agent = Agent.create(
    # Agent + Environment
    agent='ppo', environment=environment, max_episode_timesteps=num_actuations,
    # TODO: num_actuations could be specified by Environment.max_episode_timesteps() if it makes sense...
    # Network
    network=network,
    # Optimization
    batch_size=20, learning_rate=decaying_lr, subsampling_fraction=0.2, optimization_steps=25,
    # Reward estimation
    likelihood_ratio_clipping=0.2,
    # TODO: gae_lambda=0.97 doesn't currently exist
    # Critic
    critic_network=network,
    critic_optimizer=dict(
        type='multi_step', num_steps=5,
        optimizer=dict(type='adam', learning_rate=1e-3)
    ),
    # Regularization
    entropy_regularization=0.01,
    # TensorFlow etc
    saver=dict(directory=os.path.join(os.getcwd(), 'saver_data'), frequency=1200),  # the high value of the seconds parameter here is so that no erase of best_model
)

agent.initialize()

runner = Runner(agent=agent, environment=environment, save_best_agent=True)
```

I think having a strong default, without too much options would be good as it makes life simpler for the users and for you.

Do you think it would be possible to 'simply' save both, with the structure:

```
saver_data
  |- some files for the periodic saver
  | ...
  |--- best_model
            |- some files for the best_model
            |- ...
```

**Comment by AlexKuhnle** [comment]
Yes, a `best_model` subdirectory sounds like a good solution if the periodic saver is turned on.

**Comment by jerabaul29** [comment]
Btw, I have a single runner like that:

```
...

agent.initialize()
agent.load('saver_data', 'best_model')

...
```

with:

```
jrlab@jrlab-ThinkPad-T490:~/Desktop/Current/training_Re1000_no_Ccost$ ls saver_data -l
total 232868
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 13:24 agent-21695.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 13:24 agent-21695.index
-rw-r--r-- 1 jrlab jrlab 10063536 nov.   6 13:24 agent-21695.meta
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 13:44 agent-23821.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 13:44 agent-23821.index
-rw-r--r-- 1 jrlab jrlab 10063536 nov.   6 13:44 agent-23821.meta
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 14:04 agent-25944.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 14:04 agent-25944.index
-rw-r--r-- 1 jrlab jrlab 10063536 nov.   6 14:04 agent-25944.meta
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 14:24 agent-28215.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 14:24 agent-28215.index
-rw-r--r-- 1 jrlab jrlab 10063536 nov.   6 14:24 agent-28215.meta
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 14:41 agent-28800.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 14:41 agent-28800.index
-rw-r--r-- 1 jrlab jrlab 10063532 nov.   6 14:41 agent-28800.meta
-rw-r--r-- 1 jrlab jrlab      978 nov.   6 14:41 agent.json
-rw-r--r-- 1 jrlab jrlab 13777284 nov.   6 14:30 best-model.data-00000-of-00001
-rw-r--r-- 1 jrlab jrlab     2403 nov.   6 14:30 best-model.index
-rw-r--r-- 1 jrlab jrlab      978 nov.   6 14:30 best-model.json
-rw-r--r-- 1 jrlab jrlab 10063536 nov.   6 14:30 best-model.meta
-rw-r--r-- 1 jrlab jrlab      207 nov.   6 14:41 checkpoint
-rw-r--r-- 1 jrlab jrlab 38519295 nov.   6 14:24 events.out.tfevents.1573031056.jrlab-ThinkPad-T490
-rw-r--r-- 1 jrlab jrlab 38517938 nov.   6 14:41 events.out.tfevents.1573047658.jrlab-ThinkPad-T490
-rw-r--r-- 1 jrlab jrlab 18320724 nov.   6 14:41 graph.pbtxt

```

Any idea why I get this error message?

```
Traceback (most recent call last):
  File "single_runner_nosocket.py", line 56, in <module>
    agent.load('saver_data', 'best_model')
  File "/home/jrlab/.local/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 123, in load
    agent = Agent.create(agent=agent)
  File "/home/jrlab/.local/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 78, in create
    library = importlib.import_module(name=library_name)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'saver_data/best_model'

```

**Comment by MariusHolm** [comment]
I'm currently meeting quite similar issues as @jerabaul29. Some of my simulations needs to run over several days, which then risks overwriting best_model if I don't define frequency as very large. 
Would be great to have a solution as proposed above with a subdirectory for best_model saves.

**Comment by AlexKuhnle** [comment]
Will look into that very soon, maybe today.

**Comment by AlexKuhnle** [comment]
Hey both, I addressed the problem now as follows:

- `save_best_agent` can either be `False` or `True`, or a string, in which case it is used as a separate best-model saver directory.
- `True` is only allowed if the agent config dict, but not the `Agent` object is passed to the runner (in which case it is created internally), so that the saver configuration can be adapted accordingly (see above).
- I decided for now that the solution with a best-model subdirectory is a bit too "messy", since it on the one hand relies on using the TensorFlow saver object's directory setting, but on the other hand doesn't use that directory but a modified version of it instead. I think it's best to assume that TensorFlow "organizes" the saver directory.

Hope it helps, let me know if not!
